packages/daisyui/src/widgets/DateTimeWidget/DateTimeWidget.tsx

```diff
@@ -0,0 +1,325 @@
+      e.stopPropagation();
+      setIsOpen((prev) => !prev);
+      if (!isOpen && onFocus) {
+        onFocus(id, value);
+      }
+    },
+    [isOpen, id, onFocus, setIsOpen, value]
+  );
+
+  // Handle focus event
+  const handleFocus = useCallback(() => {
+    if (onFocus) {
+      onFocus(id, value);
+    }
+  }, [id, onFocus, value]);
+
+  // Handle blur event
+  const handleBlur = useCallback(() => {
+    if (!isOpen && onBlur) {
+      onBlur(id, value);
+    }
+  }, [id, onBlur, value, isOpen]);
+
+  // Close popup on escape key
+  useEffect(() => {
+    const handleEscape = (e: KeyboardEvent) => {
+      if (e.key === 'Escape' && isOpen) {
+        setIsOpen(false);
+        if (onBlur) {
+          onBlur(id, value);
+        }
+      }
+    };
+
+    document.addEventListener('keydown', handleEscape);
+    return () => document.removeEventListener('keydown', handleEscape);
+  }, [id, isOpen, onBlur, value]);
+
+  return (
+    <div className='form-control my-4 w-full relative'>
+      <div
+        className='w-full'
+        tabIndex={0}
+        onKeyDown={(e) => {
+          if (e.key === 'Enter' || e.key === ' ') {
+            togglePicker(e as unknown as React.MouseEvent);
+          }
+        }}
+        onFocus={handleFocus}
+        onBlur={handleBlur}
+        ref={inputRef}
+      >
+        <div
+          id={id}
+          className={`input input-bordered w-full flex items-center justify-between cursor-pointer ${
+            isOpen ? 'ring-2 ring-primary/50' : ''
+          }`}
+          onClick={togglePicker}
+          role='button'
+          aria-haspopup='true'
+          aria-expanded={isOpen}
+          tabIndex={-1}
+        >
+          <span className={localDate && isValid(localDate) ? '' : 'text-base-content/50'}>
+            {localDate && isValid(localDate) ? format(localDate, 'PP p') : schema.title}
+          </span>
+          <FontAwesomeIcon icon={faCalendar} className='ml-2 h-4 w-4 text-primary' />
+        </div>
+        {isOpen && (
+          <div
+            ref={containerRef}
+            className='absolute z-[100] mt-2 w-full max-w-xs bg-base-100 border border-base-300 shadow-lg rounded-box'
+            onClick={(e) => e.stopPropagation()}
+          >
+            <MemoizedDateTimePickerPopup
+              selectedDate={localDate}
+              month={month}
+              onMonthChange={handleMonthChange}
+              onSelect={handleSelect}
+              onTimeChange={handleTimeChange}
+            />
+            <div className='p-3 flex justify-end border-t border-base-300'>
+              <button
+                type='button'
+                className='btn btn-sm btn-primary'
+                onClick={() => {
+                  setIsOpen(false);
+                  onChange(localDate ? localDate.toISOString() : '');
+                  if (onBlur) {
+                    onBlur(id, value);
+                  }
+                  inputRef.current?.focus();
+                }}
```

Comment:
Consider moving complex callbacks into an external function, perhaps using `useCallback()` to avoid rerenders